controllers/user.js

```diff
@@ -0,0 +1,136 @@
+const User = require("../models/User");
+const bcrypt = require("bcrypt");
+
+module.exports = {
+  /* CREATE */
+  registerUser: async (request, response) => {
+    const {username, firstName, lastName, email, password} = request.body;
+
+    if (!password || password.length < 3) {
+      return response.status(400).json({
+        error: "`password` is shorter than the minimum allowed length (3)",
+      });
+    }
+
+    const saltRounds = 10;
+    const passwordHash = await bcrypt.hash(password, saltRounds);
+
+    const user = new User({
+      username,
+      firstName,
+      lastName,
+      passwordHash,
+      email,
+    });
+
+    const savedUser = await user.save();
+
+    response.status(201).json(savedUser);
+  },
+
+  /* READ */
+  getUsers: async (request, response) => {
+    try {
+      const user = await User.find({});
+      response.status(200).json(user);
+    } catch (err) {
+      response.status(404).json({message: err.message});
+    }
+  },
+
+  getUser: async (request, response) => {
+    try {
+      const user = await User.findById(request.params.id)
+        .populate("recipes", {
+          title: 1,
+          picturePath: 1,
+        })
+        .populate("reviews", {
+          recipeId: 1,
+          timesMade: 1,
+          rating: 1,
+          picturePath: 1,
+        })
+        .populate("user", {following: 1, followers: 1});
+      response.status(200).json(user);
+    } catch (err) {
+      response.status(404).json({message: err.message});
+    }
+  },
+
+  getUserFollowers: async (request, response) => {
+    try {
+      const followers = await User.findById(request.params.id).populate(
+        "user",
+        {followers: 1}
+      );
+      response.status(200).json(followers);
+    } catch (err) {
+      console.log(err);
+    }
+  },
+
+  getUserFollowings: async (request, response) => {
+    try {
+      const following = await User.findById(request.params.id).populate(
+        "user",
+        {following: 1}
+      );
+      response.status(200).json(following);
+    } catch (err) {
+      console.log(err);
+    }
+  },
+
+  /* UPDATE */
+  updateUser: async (request, response) => {
+    try {
+    } catch (err) {
+      console.log(err);
+    }
+  },
+
+  addRemoveFollowing: async (request, response) => {
```

Comment:
I'm really sure what this method is doing. It seems like you're trying to merge 2 methods together. A good method should be doing one thing well. I suggest you to break it up.